### ISSUE TITLE
Using LambdaCDM with emcee causes RAM to fill up very quickly

### ISSUE THREAD

**Issue by Oedon** [astropy/astropy#6354]
I'm testing out how well astropy's own `angular_diameter_distance` works for Markov Chain inference with the `emcee` module since I'm having trouble putting it in my model myself, and upon using it (or any cosmology distance function) I noticed that my RAM was filling up very quickly, much faster than when I sample without an `astropy` function involved. Here is the relevant code:

```
import emcee as em
import numpy as np
import pandas as pd
from astropy.cosmology import LambdaCDM

d1 = pd.read_csv("betterdata.csv")
M_true =  0.28458426
L_true =  0.72886156
H_true = 70.01235027
c = 299792.458
z    = np.array(d1['z'])
y    = np.array(d1['mu'])
yerr = np.array(d1['dmu'])

def lnprior(theta):
    M, H, L = theta
    if (0. < M < 2. and  50. < H < 100. and 0. < L < 2.):
        return 0.0
    else:
        return -np.inf

def lnlike(theta, z, y, yerr):
    M, H, L = theta
    cosmo = LambdaCDM(H0 = H, Om0 = M, Ode0 = L)

    model = 5. * np.log10((1.+z)**2 * cosmo.angular_diameter_distance(z).value) + 25.

    return np.sum(-0.5 * ((y-model)**2/(yerr**2)) - np.log(yerr*np.sqrt(2.*np.pi)))


def lnprob(theta, z, y, yerr):
    lp = lnprior(theta)
    if not np.isfinite(lp):
        return -np.inf
    ll = lnlike(theta, z, y, yerr)
    if not np.isfinite(ll):
        return -np.inf
    return ll + lp
    

ndim, nwalkers, nsteps = 3, 6, 200    # _minimum amount of walkers is 2*ndim for demonstration purposes_
p0 = [np.array([M_true, H_true, L_true]) + 1e-2 * np.random.randn(ndim) for i in range(nwalkers)]
sampler = em.EnsembleSampler(nwalkers, ndim, lnprob, a = 4., args=(z, y, yerr), threads = 4)

sampler.run_mcmc(p0, nsteps)
```

The sampler runs perfectly fine, but again the RAM fills up very quickly and I have no idea why. I believe that `emcee` works by calling `lnprob` for each MCMC sample, so the only thing I can think of that's causing the RAM issue is that calling `cosmo = LambdaCDM(stuff)` puts something in the memory that isn't leaving, despite it not being used after a new parameter coordinate is chosen. This same thing happens when I change my model to use `cosmo.comoving_distance(z).value` as well, though for that I can use my own calculation instead.

If this issue is unfixable due to something in `astropy`, is there a working python calculation to determine Angular Diameter Distance that doesn't use something like `astropy`? I've tried `cosmolopy` and that has the same problems.

**Comment by aconley** [comment]
I tried running your example (python 3.6, up to date deps, MacOs 10.12.5) for 
2500 randomly generated zs (0, 1], and had absolutely no memory pressure to
speak of.  So -- how big is that input file?

In general, though, that's the trouble with garbage collected languages, probably
not anything very specific to astropy.  Especially when calling out to some
external library in another language, which is where the integrals are actually
happening (scipy).

You could try explicitly doing `del cosmo` after you evaluate, although I would be 
surprised if that really helped much.

If you are willing to get really ugly, you can reuse the inner guts of the same
cosmology object over and over instead of re-allocating it on each call.  This is
evil and fragile -- meaning it may suddenly stop working in some future
astropy release, and we won't warn you -- so only if you are desperate.

Make cosmo a global object that is reused by each call and:

```
cosmo._H0 = <value> * u.km / u.Mpc / u.s
cosmo._Om0 = <value>
cosmo._Ode0 = <value>
cosmo._Ok0 = 1.0 - cosmo._Om0 - cosmo._Ode0
cosmo._hubble_distance = (const.c / self._H0).to(u.Mpc)
```

inside your prob calls. 

PS: `cosmo.distmod(z).value`

**Comment by Oedon** [comment]
I'm very sorry that I never specified my OS and Python version, that was a big oversight on my part. I'm using Python 2.7 on Ubuntu 16.04. I'm only using 2.7 because I thought I was going to be using ROOT (wrapped with PyROOT) for this project, but there were too many complications along the way to make it viable (and I'm also a bit afraid of writing pseudo-C code). For future projects I'll definitely be using Python3 but I'm stuck with 2 for now.

My `z` array is of size 580 (as well as my `y` and `yerr` arrays), so the issue is definitely not coming from that. As for `del cosmo`, I've tried both that and `gc.collect()` in a bunch of different places within and outside of `lnlike` and neither worked, with `gc.collect()` only slowing the sampling down.

Just to clarify the last bit of your response, would I put this:
```
cosmo._H0 = <value> * u.km / u.Mpc / u.s
cosmo._Om0 = <value>
cosmo._Ode0 = <value>
cosmo._Ok0 = 1.0 - cosmo._Om0 - cosmo._Ode0
cosmo._hubble_distance = (const.c / self._H0).to(u.Mpc)
```
in `lnlike` verbatim or put it in a function that returns the `angular_diameter_distance` array?

**Comment by aconley** [comment]
It's up to you where you put it, although encapsulating it in a function is probably good design.

```
import astropy.units as u
import astropy.constants as const

.. omit a bunch of stuff ..

cosmo = LambdaCDM(H0=70, Om0=0.3, Ode0=0.7)

def get_distance_modulus(z, theta):
  Om0, H0, Ode0 = theta
  cosmo._H0 = H0 * u.km / u.Mpc / u.s
  cosmo._hubble_distance = (const.c / self._H0).to(u.Mpc)
  cosmo._Om0 = Om0
  cosmo._Ode0 = Ode0
  cosmo._Ok0 = 1.0 - cosmo._Om0 - cosmo._Ode0
  return cosmo.distmod(z).value

def lnlike(theta, z, y, yerr):
  model = get_distance_modulus(z, theta)
  return np.sum(-0.5 * ((y-model)**2/(yerr**2)) - np.log(yerr*np.sqrt(2.*np.pi)))
```

or something close to that should work.  Although it does prevent you from
running emcee in parallel because of the race conditions in the update to cosmo.

**Comment by Oedon** [comment]
So I tried that, and upon trying to run it, the interpreter had an issue with:
```
---> 13     cosmo._hubble_distance = (const.c / self._H0).to(u.Mpc)

NameError: global name 'self' is not defined
```
I changed that line to read `cosmo._H0` instead of `self._H0`, which made it run but I still had that memory issue. That sounds like a Python2 specific error?

In any case, after a lot of searching I found the [`cosmology`](https://github.com/esheldon/cosmology) module which can also calculate angular diameter distance, and it works very well without any of the memory issues. Not sure why at all, though. Thanks for the help regardless!
